docs(auth): add example on returning custom error messages on auth handlers (#79)
encoredev/encore: docs/develop/auth.md

}
```

## Handling auth errors

When a token doesn't match your auth rules (for example if it's expired, the token has been revoked, or the token is invalid), you should return a non-nil error from the auth handler.

Encore passes the error message on to the user when you use [Encore's built-in error package](errors), so we recommend using that with the error code `Unauthenticated` to communicate what happened. For example:

```go

> import (

Reviewer: I think this code sample can drop the imports and the `// AuthHandler can be named whatever you prefer (but must be exported).` since the previous code samples cover these.

Author: Okay

Author: `encore.dev/beta/errs` is not imported by other code samples ?

}
```

> ## Returning custom error messages on auth handlers

Reviewer: This feels too specific; I think what we want to communicate is more along the lines of "here's how to signal that the request isn't authenticated".

How about:

```suggestion
## Handling auth errors
```

> When you want to return a custom error message in the auth handler e.g To return an `unauthenticated` error message:

Reviewer: In the same vein as above, I think we should set the stage for the reader and give a bit more detail:

```suggestion
When a token doesn't match your auth rules (for example if it's expired, the token has been revoked, or the token is invalid), you should return a non-nil error from the auth handler.

Encore passes the error message on to the user when you use [Encore's built-in error package](errors), so we recommend using that with the error code `Unauthenticated` to communicate what happened. For example: 
```

```go

> import (
>     "encore.dev/beta/auth"
>     "encore.dev/beta/errs"
> )
> 
> // AuthHandler can be named whatever you prefer (but must be exported).
> //encore:authhandler
> func AuthHandler(ctx context.Context, token string) (auth.UID, error) {
>     // if the token is invalid/expired you can return a custom error message
>     // using the code below.
>     return "", errs.B().Code(errs.Unauthenticated).Msg("auth token is invalid").Err()
> }

Reviewer: I think we can simplify this example slightly. How about:

```suggestion
//encore:authhandler
func AuthHandler(ctx context.Context, token string) (auth.UID, error) {
    return "", &errs.Error{
        Code: errs.Unauthenticated,
        Message: "invalid token",
    }
}
```

```

> 

Reviewer: Finally it would be great to add a callout on security. Something like this:

```suggestion
<Callout type="important">

Note that for security reasons you may not want to reveal too much information about why a request did not pass your auth checks. There are many subtle security considerations when dealing with authentication, and we don't have time to go into all of them here. Whenever possible we recommend using a third-party auth provider; see [Using Firebase Authentication](../how-to/firebase-auth) for an example of how to do that.

</Callout>

Author: @eandre i just pushed a commit to improve on the changes

}
```

## Handling auth errors

When a token doesn't match your auth rules (for example if it's expired, the token has been revoked, or the token is invalid), you should return a non-nil error from the auth handler.

Encore passes the error message on to the user when you use [Encore's built-in error package](errors), so we recommend using that with the error code `Unauthenticated` to communicate what happened. For example:

```go
import "encore.dev/beta/errs"

//encore:authhandler
func AuthHandler(ctx context.Context, token string) (auth.UID, error) {
    return "", &errs.Error{
        Code: errs.Unauthenticated,
        Message: "invalid token",
    }
}

> 

Reviewer: Remove these two empty lines; they're part of a code block so would show up in the fully rendered doc.

Author: @eandre i have removed the empty lines

// AuthHandler can be named whatever you prefer (but must be exported).
//encore:authhandler
func AuthHandler(ctx context.Context, token string) (auth.UID, error) {
    return "", &errs.Error{
        Code: errs.Unauthenticated,
        Message: "invalid token",
    }
}

```

<Callout type="important">

> Note that for security reasons you may not want to reveal too much information about why a request did not pass your auth checks. There are many subtle security considerations when dealing with authentication and we don't have time to go into all of them here. Whenever possible we recommend using a third-party auth provider; see [Using Firebase Authentication](../how-to/firebase-auth) for an example of how to do that.

Reviewer: There needs to be a newline between `<Callout type="important">` and the content (and similarly between the content and the closing tag `</Callout>`) for MDX parsing reasons. Don't ask me why :)

Author: Okay, noted

</Callout>

> 

Reviewer: Minor nitpick: can remove the extra newline here

Author: Okay
